### DISCUSSION TITLE
How to clip things drawn in custom Contrel.Render() ?

### DISCUSSION THREAD

**Discussion by sonicsmooth** [AvaloniaUI/Avalonia#9774] [Q&A] [answered]
Avalonia beginner here.

I have a barebones custom control which just draws some stuff in `public override void Render(...)`.
The contents are drawn basically to the edges of `this.Bounds`.
However when I apply a rotation or scale transform, the contents are drawn outside the bounds, such as if placed in a `Grid` layout; the contents spill over into the next grid cells.
I'm able to place this control in a `Panel` with `ClipToBounds="True"`, but I'd prefer not to have to nest this inside a `Panel`.  I'm also not sure how to place this in a `Panel` without having to create two .xaml files (an outer and inner file).

1. Is there a way to get my custom control to utilize some built-in clipping functionality without having to place this inside a `Panel`?
2. If I do have to place my custom control in a `Panel` for its clipping, can I do this with one .xaml file and one class?
3. Deriving from `Panel` with `ClipToBounds="True"` instead of deriving from `Control` doesn't work.

This is what it looks like without clipping.
![image](https://user-images.githubusercontent.com/672429/209425847-bfca8400-17ef-4b25-9084-f9076500442e.png)

Thanks.

**Comment by kekekeks** [comment, marked as answer]
Applying a render transform to a control transforms it's bounds too. So ClipToBounds still works but clips to rotated bounds.

**Reply by sonicsmooth** [reply]
Thanks for helping me understand.  What is the best solution? Can I add another clipping operation with its own (unrotated) transform in the same object?  Or do something like render the rotated graphics, then unrotate only the clipping operation and apply it again?  etc.

**Reply by kekekeks** [reply]
Do you need to rotate the entire control? You can push rotate transform from your Render method.

**Reply by sonicsmooth** [reply]
Yes, the entire control (just a grid) needs to be able to rotate and scale, and clip to the bounds given by its container.  I'll check into push rotate transform.  Thanks.

**Reply by sonicsmooth** [reply]
I figured out to use 
    `using (context.PushPostTransform(new RotateTransform(Rotation).Value)) { /* draw stuff */ }`.  Now the grid is drawn inside the proper bounds.

thanks!
